Products/CMFQuickInstallerTool/QuickInstallerTool.py

```diff
@@ -817,5 +846,94 @@ def getInstanceHome(self):
+                newVersion=product_version,
+            )
+        profile_id = profile['id']
+        setup = getToolByName(self, 'portal_setup')
+        profile_version = str(setup.getVersionForProfile(profile_id))
+        if profile_version == 'latest':
+            profile_version = self.getLatestUpgradeStep(profile_id)
+        if profile_version == 'unknown':
+            # If a profile doesn't have a metadata.xml use product version
+            profile_version = product_version
+        installed_profile_version = setup.getLastVersionForProfile(profile_id)
+        # getLastVersionForProfile returns the version as a tuple or unknown.
+        if installed_profile_version != 'unknown':
+            installed_profile_version = str(
+                '.'.join(installed_profile_version))
+        return dict(
+            required=profile_version != installed_profile_version,
+            available=len(setup.listUpgrades(profile_id)) > 0,
+            hasProfile=True,
+            installedVersion=installed_profile_version,
+            newVersion=profile_version,
+        )
+
+    @security.protected(ManagePortal)
+    def getLatestUpgradeStep(self, profile_id):
+        """
+        Get the highest ordered upgrade step available to
+        a specific profile.
+
+        If anything errors out then go back to "old way"
+        by returning 'unknown'
+        """
+        setup = getToolByName(self, 'portal_setup')
+        profile_version = 'unknown'
+        try:
+            available = setup.listUpgrades(profile_id, True)
+            if available:  # could return empty sequence
+                latest = available[-1]
+                profile_version = max(latest['dest'],
+                                      key=pkg_resources.parse_version)
+        except Exception:
+            pass
+
+        return profile_version
+
+    @security.protected(ManagePortal)
+    def upgradeProduct(self, pid):
+        profile = self.getInstallProfile(pid)
+        if profile is None:
+            # No upgrade profiles
+            return self.reinstallProducts(products=[pid])
+        profile_id = profile['id']
+        setup = getToolByName(self, 'portal_setup')
+        upgrades = setup.listUpgrades(profile_id)
+        for upgrade in upgrades:
+            # An upgrade may be a single step (for a bare upgradeStep)
+            # or a list of steps (for upgradeSteps containing upgradeStep
+            # directives).
+            if not isinstance(upgrade, list):
+                upgrade = [upgrade]
+            for upgradestep in upgrade:
+                step = upgradestep['step']
+                step.doStep(setup)
+        version = str(profile['version'])
+        if version == 'latest':
+            version = self.getLatestUpgradeStep(profile_id)
+        setup.setLastVersionForProfile(profile_id, version)
+
 
 InitializeClass(QuickInstallerTool)
+# TODO: we probably want this, copied from CMPlone:
```

Comment:
@mauritsvanrees I usually tend to add TODOs as `TODO(gforcada): lorem ipsum...`. Question though, when would it be a good time to decide if we want to copy it or not? 🤔

Reply:
Adding a name to a `TODO` comment is a good idea.

Actually, I see it is useless to call `registerToolInterface` here. When we register it, it should mean that a `getToolByName(context, 'portal_quickinstaller')` would try to get the `IQuickInstallerTool` utility first, and fall back to getting the tool. But even in current Plone, there is no such utility registered in zcml. I briefly tried to do that, but then Plone failed to start.
Even if the utility call *would* work, you would probably get a utility that does not really work: the QI *really* needs to be a tool, storing stuff in the database, especially `InstalledProducts`.

I have removed the commented out lines now.